frontend/src/components/features/github/github-repo-selector.tsx

```diff
@@ -16,25 +34,43 @@ export function GitHubRepositorySelector({
 }: GitHubRepositorySelectorProps) {
   const { data: config } = useConfig();
   const [selectedKey, setSelectedKey] = React.useState<string | null>(null);
+  const [searchQuery, setSearchQuery] = React.useState<string>("");
+  const [searchedRepos, setSearchedRepos] = React.useState<
+    GitHubRepositoryWithFlag[]
+  >([]);
+  const debouncedSearchQuery = useDebounce(searchQuery, 300);
 
-  // Add option to install app onto more repos
-  const finalRepositories =
-    config?.APP_MODE === "saas"
-      ? [{ id: -1000, full_name: "Add more repositories..." }, ...repositories]
-      : repositories;
+  React.useEffect(() => {
+    const searchPublicRepo = async () => {
+      if (!debouncedSearchQuery) {
+        setSearchedRepos([]);
+        return;
+      }
+      const repos = await searchPublicRepositories(
+        sanitizeQuery(debouncedSearchQuery),
+      );
+      // Sort by stars in descending order
+      const sortedRepos = repos
+        .sort((a, b) => (b.stargazers_count || 0) - (a.stargazers_count || 0))
+        .slice(0, 3);
+      setSearchedRepos(sortedRepos);
+    };
+
+    searchPublicRepo();
+  }, [debouncedSearchQuery]);
+
+  const finalRepositories: GitHubRepositoryWithFlag[] = [
+    ...searchedRepos.filter(
+      (repo) => !repositories.find((r) => r.id === repo.id),
+    ),
+    ...repositories,
+  ];
```

Comment:
We can outsource this into a hook that utilizes TanStack Query. That way we can control when the query is executed (`!!debouncedSearchQuery`) and return a sorted array (via the `select` property of the query/mutation hook).

That should help us get rid of the `useEffect` and `useState` (`searchedRepos`) and make the component smaller and more readable

Reply:
The repository search logic has been refactored to use TanStack Query. The useSearchRepositories hook now handles the query execution, sorting, and error handling. The component is now cleaner and more maintainable. (by OpenHands)